File: AlgorithmsPackJavaSolutions/1591. Strange Printer II.java
class Solution {
    private List<List<Integer>> graph;
    
    private void addEdge(List<List<Integer>> graph, int from, int to) {
        graph.get(from).add(to);
    }
    
    private List<Integer> topSort(List<List<Integer>> graph) {
        int size = graph.size();
        int[] incomingEdges = new int[size];
        for (List<Integer> neighbors : graph) {
            for (int neighbor : neighbors) {
                incomingEdges[neighbor]++;
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < size; i++) {
            if (incomingEdges[i] == 0) {
                queue.add(i);
            }
        }
        
        List<Integer> ordering = new ArrayList<>();
        
        while (!queue.isEmpty()) {
            int i = queue.poll();
            ordering.add(i);
            
            for (int j : graph.get(i)) {
                if (--incomingEdges[j] == 0) {
                    queue.add(j);
                }
            }
        }
        
        if (ordering.size() != size) {
            ordering.clear();
        }
        
        return ordering;
    }
    
    public boolean isPrintable(int[][] targetGrid) {
        int rows = targetGrid.length;
        int cols = targetGrid[0].length;
        int maxColor = 61;
        
        int[] rmax = new int[maxColor];
        int[] rmin = new int[maxColor];
        int[] cmax = new int[maxColor];
        int[] cmin = new int[maxColor];
        
        Arrays.fill(rmax, -1);
        Arrays.fill(rmin, rows);
        Arrays.fill(cmax, -1);
        Arrays.fill(cmin, cols);
        
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                int color = targetGrid[r][c];
                rmax[color] = Math.max(rmax[color], r);
                rmin[color] = Math.min(rmin[color], r);
                cmax[color] = Math.max(cmax[color], c);
                cmin[color] = Math.min(cmin[color], c);
            }
        }
        
        graph = new ArrayList<>();
        for (int i = 0; i < maxColor; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int color = 0; color < maxColor; color++) {
            for (int r = rmin[color]; r <= rmax[color]; r++) {
                for (int c = cmin[color]; c <= cmax[color]; c++) {
                    if (targetGrid[r][c] != color) {
                        addEdge(graph, color, targetGrid[r][c]);
                    }
                }
            }
        }
        
        return topSort(graph).size() > 0;
    }
}

File: AlgorithmsPackPythonSolutions/1591. Strange Printer II.py
class Solution:
    def isPrintable(self, targetGrid: List[List[int]]) -> bool:
        rows = len(targetGrid)
        cols = len(targetGrid[0])
        maxColor = 61
        
        rmax = [-1] * maxColor
        rmin = [rows] * maxColor
        cmax = [-1] * maxColor
        cmin = [cols] * maxColor
        
        for r in range(rows):
            for c in range(cols):
                color = targetGrid[r][c]
                rmax[color] = max(rmax[color], r)
                rmin[color] = min(rmin[color], r)
                cmax[color] = max(cmax[color], c)
                cmin[color] = min(cmin[color], c)
        
        graph = [[] for _ in range(maxColor)]
        
        for color in range(maxColor):
            for r in range(rmin[color], rmax[color] + 1):
                for c in range(cmin[color], cmax[color] + 1):
                    if targetGrid[r][c] != color:
                        graph[color].append(targetGrid[r][c])
        
        return len(self.topSort(graph)) > 0
    
    def topSort(self, graph: List[List[int]]) -> List[int]:
        sz = len(graph)
        incomingEdges = [0] * sz
        for i in range(sz):
            for j in graph[i]:
                incomingEdges[j] += 1
        
        queue = deque()
        ordering = []
        
        for i in range(sz):
            if incomingEdges[i] == 0:
                queue.append(i)
        
        while queue:
            i = queue.popleft()
            ordering.append(i)
            
            for j in graph[i]:
                incomingEdges[j] -= 1
                if incomingEdges[j] == 0:
                    queue.append(j)
        
        if len(ordering) != sz:
            ordering.clear()
        
        return ordering
